crates/oxc_linter/src/rules/oxc/no_redundant_constructor_init.rs

```diff
@@ -58,53 +58,48 @@ impl Rule for NoRedundantConstructorInit {
         if method.kind != MethodDefinitionKind::Constructor {
             return;
         }
-        let public_members = method
-            .value
-            .params
-            .items
-            .iter()
-            .filter_map(
-                |param| if param.is_public() { param.pattern.get_identifier_name() } else { None },
-            )
-            .collect::<Vec<_>>();
-        if public_members.is_empty() {
-            return;
-        }
-        let Some(body) = method.value.body.as_ref() else {
-            return;
-        };
-        body.statements.iter().for_each(|stmt| {
-            let Statement::ExpressionStatement(expr_stmt) = stmt else {
-                return;
-            };
-            let Expression::AssignmentExpression(assignment_expr) = &expr_stmt.expression else {
-                return;
+        for param in &method.value.params.items {
+            if !param.is_public() {
+                continue;
+            }
+            let Some(ident) = param.pattern.get_binding_identifier() else {
+                continue;
             };
+            for reference in ctx.symbol_references(ident.symbol_id()) {
+                // check if the param is being read which would indicate an assignment
+                if !reference.is_read() {
+                    continue;
+                }
 
-            // check for assigning to this: this.x = ?
-            let AssignmentTarget::StaticMemberExpression(static_member_expr) =
-                &assignment_expr.left
-            else {
-                return;
-            };
-            let Expression::ThisExpression(_this_expr) = &static_member_expr.object else {
-                return;
-            };
-            let assignment_name = static_member_expr.property.name;
+                let Some(parent_node) = ctx.nodes().parent_node(reference.node_id()) else {
+                    continue;
+                };
+                let AstKind::AssignmentExpression(assignment_expr) = parent_node.kind() else {
+                    continue;
+                };
 
-            // check both sides of assignment have the same name: this.x = x
-            let Expression::Identifier(assignment_target_ident) = &assignment_expr.right else {
-                return;
-            };
-            if assignment_target_ident.name != assignment_name {
-                return;
-            }
+                // check for assigning to this: this.x = ?
+                let AssignmentTarget::StaticMemberExpression(static_member_expr) =
+                    &assignment_expr.left
+                else {
+                    continue;
+                };
+                let Expression::ThisExpression(_this_expr) = &static_member_expr.object else {
+                    continue;
+                };
```

Comment:
if you don't need `_this_expr`:

```suggestion
                if !matches!(&static_member_expr.object, Expression::ThisExpression(_)) {
                    continue;
                }
                let Expression::ThisExpression(_this_expr) = &static_member_expr.object else {
                    continue;
                };
```